runtime/lua/vim/treesitter/query.lua

```diff
@@ -249,7 +249,32 @@ local predicate_handlers = {
     end
 
     return false
-  end
+  end,
+
+  ["any-of?"] = (function()
+    local sets = {}
+
+    return function(match, _, source, predicate)
+      local node = match[predicate[2]]
+      local node_text = M.get_node_text(node, source)
+      local start_row, _, end_row, _ = node:range()
+      if start_row ~= end_row then
+        return false
+      end
+
+      local k = predicate
+      local set = sets[k]
```

Comment:
Maybe instead of using this `sets` variable, you could associate the cache directly to the predicate?

As the predicate objects are created on query creation, you would not lose you cache.

Reply:
But one would need a marker entry to see whether a table miss is because it's not in the table or whether the look-up has not been constructed yet. So it would be pretty similar to the set implementation right now.

Reply:
Is there a particular implementation you are thinking of? I can think two:

1. Something like `predicate["cache"] = {}`, but this would mess up the `#` operator.
2. Store it in a table of the metatable of predicate, but this would prevent `predicate` from having an official metatable in the future.

Reply:
I completely agree. It's probably better to have a separate table for it as it's right.i suppose  # should still work. Doesn't it only consider numerical indices?

Reply:
I'm not sure. Would having non-integer keys in a table cause undefined behavior when using `#`?

Reply:
It could be defined... But I see no need to store directly to `predicate`
```
The length of a table t is defined to be any integer index n such that t[n] is not nil and t[n+1] is nil; moreover, if t[1] is nil, n can be zero. For a regular array, with non-nil values from 1 to a given n, its length is exactly that n, the index of its last value. If the array has "holes" (that is, nil values between other non-nil values), then #t can be any of the indices that directly precedes a nil value (that is, it may consider any such nil value as the end of the array). 
```

Reply:
It feels kleenr to me because predicates id's are not guaranteed to be uniques between queries.
Let's say I have a js query, with an `any-of? ` predicate, associated with id 15, and the same for a scheme query, then those two would collapse and we have no clue on which would win.

Note two different queries with the same language would make the same issue, so associating the language would not help, we **need** to associate this cache to the predicate itself rather than a global thing.

Reply:
But we're using the hash of `predicates` which is the same as assuming that `predicates` is the same to store stuff into it.

Reply:
either the present solution or @vigoux :s alternative should work. The only difference I see is that vigoux suggestion is easier to introspect, potentially.

Reply:
> But we're using the hash of `predicates` (not the ID of the predicate) which is the same as assuming that `predicates` is the same to store stuff into it. 

OH yeah you're right sorry.

But in a'u case there's something that does not feel right, and the `#` operator is a non-issue, because the only moment where we are going to manipulate the predicate is in the handler, so we could just specialize the handler for this case anyway.

Reply:
a'u?